src/get-next-connection-state.ts

```diff
@@ -0,0 +1,141 @@
+
+export interface GetNextConnectionStateInput {
+  currConnType: ConnectionType;
+  currStatus: ConnectionStatus;
+  deviceStatus: DeviceConnectionStatus;
+  prevDeviceStatus: DeviceConnectionStatus | null;
+  type: StatusListenerType;
+  hasAttempedReconnect: boolean;
+  setHasAttemptedReconnect: (val: boolean) => void;
+  onFirstConnectAttempt: boolean;
+  setOnFirstConnectAttempt: (val: boolean) => void;
+}
+
+export type NextConnectionState =
+  | { status: ConnectionStatus; flowType: ConnectionFlowType }
+  | undefined;
+
+export const getNextConnectionState = ({
+  currConnType,
+  currStatus,
+  deviceStatus,
+  prevDeviceStatus,
+  type,
+  hasAttempedReconnect,
+  setHasAttemptedReconnect,
+  onFirstConnectAttempt,
+  setOnFirstConnectAttempt,
+}: GetNextConnectionStateInput): NextConnectionState => {
+  const flowType =
+    type === "usb"
+      ? ConnectionFlowType.RadioBridge
+      : type === "radioRemote"
+      ? ConnectionFlowType.RadioRemote
+      : ConnectionFlowType.Bluetooth;
+
+  // We are using usb status to infer the radio bridge device status.
+  if (type === "usb") {
+    // Ignore USB status updates when radio connection is not established.
+    if (currConnType !== "radio" || onFirstConnectAttempt) {
+      return undefined;
+    }
+    if (
+      !hasAttempedReconnect &&
+      currStatus === ConnectionStatus.Connected &&
+      deviceStatus === DeviceConnectionStatus.DISCONNECTED
+    ) {
+      setHasAttemptedReconnect(true);
+      return { status: ConnectionStatus.ConnectionLost, flowType };
+    }
+    if (
+      !hasAttempedReconnect &&
+      currStatus !== ConnectionStatus.Connected &&
+      deviceStatus === DeviceConnectionStatus.DISCONNECTED
+    ) {
+      setHasAttemptedReconnect(true);
+      return { status: ConnectionStatus.FailedToReconnect, flowType };
+    }
+    if (
+      hasAttempedReconnect &&
+      deviceStatus === DeviceConnectionStatus.DISCONNECTED
+    ) {
+      return {
+        status: ConnectionStatus.FailedToReconnectTwice,
+        flowType: ConnectionFlowType.RadioRemote,
+      };
+    }
+    return undefined;
+  }
+  if (
+    // Disconnection happens for newly started / restarted
+    // bluetooth connection flows when clearing device
+    deviceStatus === DeviceConnectionStatus.DISCONNECTED &&
+    currStatus === ConnectionStatus.NotConnected
+  ) {
+    return { status: ConnectionStatus.NotConnected, flowType };
+  }
+  if (deviceStatus === DeviceConnectionStatus.CONNECTED) {
+    setOnFirstConnectAttempt(false);
+    setHasAttemptedReconnect(false);
+    return { status: ConnectionStatus.Connected, flowType };
+  }
+  if (
+    onFirstConnectAttempt &&
+    currStatus === ConnectionStatus.Connecting &&
+    deviceStatus === DeviceConnectionStatus.DISCONNECTED
+  ) {
+    return { status: ConnectionStatus.FailedToConnect, flowType };
+  }
+  if (
+    // If user does not select a device for bluetooth connection
```

Comment:
I wonder if all of these could do with a single line comment like this so the next person to work on this can find the right section without having to read all the `if` statements.

Reply:
https://github.com/microbit-foundation/ml-trainer/pull/281/commits/cbbfcb5c08f1d5f2b4797a24009d7944c62169ca